### ISSUE TITLE
Save Excel Worksheet as PDF

### ISSUE THREAD

**Issue by wgshef** [dfinke/ImportExcel#613]
Is there a way to save a worksheet as a PDF file?
In Delphi, I would use:
    Sheet.ExportAsFixedFormat(0, 'FileName.pdf', EmptyParam, EmptyParam, EmptyParam, 1, 1, EmptyParam, EmptyParam);
    // Ref: https://powerspreadsheets.com/save-excel-file-pdf-vba/
    // Type: 0=PDF, 1= XPS
    // Filename
    // Quality: 0=Standard, 1=Minimum
    // IncludeDocProperties: True/False
    // IgnorePrintAreas: True/False
    // From #
    // To #
    // OpenAfterPublish: True/False
    // FixedFormatExtClassPtr

With this example:
$excel = Get-Service | Select-Object -Property Status,Name,DisplayName,StartType | Export-Excel -Path .\test.xlsx -PassThru
I don't see an appropriate method under:
$excel
$excel.Workbook
$excel.Workbook.Worksheets[0]

Thanks!

**Comment by dfinke** [comment]
That'd be a good feature. It is not part of ImportExcel. Needs investigation. I believe you can save to PDF from the Excel application, which makes me think you could make this work if you have Excel installed and use the COM interface. These are uncharted waters.

**Comment by dfinke** [comment]
@jhoneill have you come across anything for this? I'm going to take look, thinking maybe add another open source pkg.

**Comment by jhoneill** [comment]
You can upload to office 365 and download as a PDF. I might even have that in my Graph API module. 
 
You can  use COM to do it. Open the file like this  
```
 $Excel = New-Object -ComObject "Excel.Application"
$Excel.Visible = $false
$null = $Excel.Workbooks.Open($TheFullPathToXLSX)
```
it looks like @wgshef  is calling this 
https://docs.microsoft.com/en-us/office/vba/api/excel.workbook.exportasfixedformat but
`$Excel.ActiveWorkbook.ExportAsFixedFormat(0, $TheFullPathToPDF)`
or even 
`$Excel.ActiveWorkbook.ExportAsFixedFormat(0)`
if you want to keep the name and location will work  . If you want page 1  to one it might be ,$null a few times the ,1 ,1 

then  close like this 
```
$Excel.ActiveWorkbook.Close()
$Excel.Quit()
```
Wayne , if you know how to do this in other languages is this enough to translate what's familiar into powershell ? 

You can do this but it needs user interaction to enter the name of the PDF file. 
```
[void](Get-WmiObject -Class Win32_Printer -filter "name='Microsoft Print to PDF'").SetDefaultPrinter()
start "$env:Temp\test.xlsx" -Verb "Print"
```
Obviously this and the COM route need Excel installed ,  there may be tools out there which will convert office to PDF without having office and which don't have a licence hit so can be popped onto servers. But I don't know any off the top of my head.

**Comment by wgshef** [comment]
What you had above was how to do it in COM. I was just trying to avoid using COM, but I'm using it for this.

There were a couple of printer settings that I couldn't find: Zoom, FitToPagesWide, FitToPagesTall (I do have the PaperSize and Orientation settings working, as well as the column formatting). I also couldn't find the InchesToPoints function to use with setting the Left/Right/Top/Bottom margin settings. Are these just not implemented?

Something that I've noticed: calling Export-Excel with the -AutoSize parameter does not size the columns the same as doing it from within Excel (or with the COM object's column.AutoFit() method)... when it's sized in Export-Excel the column is slightly smaller. Any idea of why?

Thanks for this great tool!

**Comment by dfinke** [comment]
Yeah, the COM approach works today. Need to find and integrate a 3rd party package to make it work without Excel. 

I believe the autosize can only approximate. Once displayed actual font sizes are in play. It does not fit long dates well at all.

**Comment by g-pearl** [comment]
For what it is worth:  I have used the following to print to PDF all sorts of documents (excel/html/csv/etc..) 

https://wkhtmltopdf.org/

**Comment by dfinke** [comment]
@g-pearl Can this take and Excel file?

**Comment by g-pearl** [comment]
I went back and looked at my code and where I thought it was using excel was an html table instead.  Sorry.  Please disregard.

**Comment by nicolasr75** [comment]
I want to add that using the COM interface has it's own problems when it comes to automation. We create reports with C#/COM/Excel and everything worked fine until we tried to automate it via Windows Task Scheduler without a logged in user. We faced all sorts of COM security troubles that were very hard to analyze. Almost every change in the security settings need a system restart. Different systems behaved differently so we did not even find a final solution. I hoped I could use ImportExcel to completely get away from the Excel COM interface. But without a solution for creating/printing PDF we may as well stay with our old code which would be sad because I like ImportExcel very much :-(
